ext/firewall.ts

```diff
@@ -0,0 +1,44 @@
+import { createExtension } from "../mod.ts";
+
+let vpnList: string[];
+let datacenterList: string[];
+let customList: string[];
+
+type FirewallOptions = {
+  blockVPN?: boolean;
+  blockDatacenter?: boolean;
+  customRange?: string[];
+  /** a list of urls that return a list of ip ranges to block in plaintext separated by newlines */
+  customLists?: string[];
+}
+
+export const firewall = createExtension<FirewallOptions>({
+  async onPlugIn({settings: opts}) {
+    await update(opts);
+    // setInterval(() => {update(opts)}, 300000);
+  },
+  onRequest({_: opts, app}) {
+    if (opts?.blockVPN && vpnList.find((range) => isIpInRange(app.ip, range))) return new Response(null, {status: 403});
+    if (opts?.blockDatacenter && datacenterList.find((range) => isIpInRange(app.ip, range))) return new Response(null, {status: 403});
+    if (opts?.customRange && opts.customRange.find((range) => isIpInRange(app.ip, range))) return new Response(null, {status: 403});
+    if (customList.find((range) => isIpInRange(app.ip, range))) return new Response(null, {status: 403});
+  }
+})
+
+async function update(args?: FirewallOptions) {
+  vpnList = (await (await fetch("https://cdn.jsdelivr.net/gh/X4BNet/lists_vpn/output/vpn/ipv4.txt")).text()).split('\n');
+  datacenterList = (await (await fetch("https://cdn.jsdelivr.net/gh/X4BNet/lists_vpn/output/datacenter/ipv4.txt")).text()).split('\n');
+  if (args?.customLists) customList = (await Promise.all(args.customLists.map((url) => fetch(url).then((res) => res.text())))).join('\n').split('\n');
```

Comment:
.

Reply:
i think the data should be fetched once, stored in a global variable (uppercase letters) with another global variable (also uppercase letters) which stores the last timestamp (in ms since epoch) since last update.

then, there could be a check whether the last update is older than 15 minutes, if so, refetch.

**regarding the fetching,** i'd suggest to use `raw.githubusercontent.com` instead of jsdelivr as this data should be more up-to-date than the one available via jsdelivr.

i'd just simplify the code, e.g. store the response in a variable, then a new variable for the data, e.g.:
```ts
const res = await fetch('...')
const data = await res.text()
```

this should improve the readability of the code.

Reply:
also, if you set the header `cf-connecting-ip`, you need to set the `proxy` option to `cloudflare` in order that cheetah uses the ip address in the header.